### DISCUSSION TITLE
Airflow freezes when using a PyTorch model

### DISCUSSION THREAD

**Discussion by kiranzo** [apache/airflow#34083] [Q&A]
Without going into excruciating detail - we have a pipeline that uses a PyTorch model to validate results. It's called within one of the Python operators. The moment it hits the line that calls neural network, execution freezes and airflow kills the unresponsive task some 12 hours later. No logs, no elaboration, no nothing, just freeze and kill.

With some debugging we could determine that the model attempts to do a forward pass, and then nothing happens. I also tried to run the code that calls the network in python console (the same python instance airflow runs on), and it worked as it should.

We honestly have no idea what's happening and how to find the source of this freeze. Did something like this happen anyone else?
I could scramble together a repository to demonstrate how the model doesn't work from within Airflow, if someone is interested in helping.

I run Airflow in Docker Desktop for Windows and trigger DAG manually, if this matters.

Thanks.

**Comment by potiuk** [comment]
> With some debugging we could determine that the model attempts to do a forward pass, and then nothing happens. I also tried to run the code that calls the network in python console (the same python instance airflow runs on), and it worked as it should.

Wild guess (and I have no experience with PyTorch - just a bit with running ML trainings in docker). There might be several reasons I can think of:

1) did you run your console as root user maybe ? If you are using GPUS, that might be the problem that teh airflow users might not have access to it (and for example fallbacks to using CPU) .

2) airflow (like any other app) runs in non-interactive mode (loolk it up if not sure). Depends how you enter the console, your console is likely interactive session - and it means that it can have some env variables set and other code executed via .bashrc for example (see https://stackoverflow.com/questions/42725538/set-environment-variables-for-non-interactive-shell for example).. Check what env variables are set in Airflow (you can dump them from Python task) comparing to your Python console.

3) Also Pytorch might ask some questions or expect interactive terminal (I do not know Pytorch so it's a wild guess) - and because airflow will run as daemon process in non-interactive mode Airflow might not provide terminal to Pytorch (and it might hang while trying to interactively ask for something)

4) Airlfow Python interpreter will  take some memory when launching task - importing some libraries etc. Could be that Pytorch simply gets stuck when the memory is not enough

**Reply by kiranzo** [reply]
Basically, we solved it by setting `torch.set_num_threads(1)` after the model initialization. There were 4 threads before, and apparently, they blocked each other when run specifically in Airflow. I still think someone with good knowledge of Torch and threading withing Airflow should take a look, this is not right.
I'm ready to provide minimal code if someone wants to tackle this.

**Reply by potiuk** [reply]
That might be billiard vs multiprocessing then (depends whether those are real threads or maybe forked multiprocessing proceeses).

If you are using Celery, then Celery uses `billiard` under-the-hood. Billiard is a drop-in replacment for multiprocessing (you need to run `import billiard as multiprocessing' if you want to use multiprocessing in Celery task) and apparently it's not compatible with standard multiprocessing,

Apparently Pytorch does the same: they have a wrapper around multiprocessing: https://pytorch.org/docs/stable/multiprocessing.html 

So quite likely this is it (if you use Celery).

I think what could help is if instead starting torch in the same proces, you spawn a new python interpreter and run your torch through that. this will not intialize Celery billiard, will be slightly slower to start and take more memory but have a chance to use more of your CPUS.

**Reply by kiranzo** [reply]
No, we are using local executor so far, this project is fairly new, and we're trying to use simpler options everywhere. Maybe we'll move to combination of Airflow and Spark in the future, I don't know yet. 

> spawn a new python interpreter 

Yeah, I considered doing this, but the way the code is right now, it's pretty hard to isolate NN usage from it. We'll have to cope with single-threading for now.

Thanks for your answers!

**Reply by potiuk** [reply]
If you are using local executor you might also try to change this:  https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#mp-start-method from `fork` to `spawn`.  Airflow ALSO uses multiprocessing for Local Executor and it forks child processes and uses multiprocessing to communicate executor <-> child processes. So maybe with the `spawn` or `forkserver` mode of multiprocessing , the problem wll be gone (`spawn`  is of course slower to start - but forking might be the cause of the multiprocessing extensions of pytorch).  

This does not require any changes in your tasks, how it will work it will start new python interpreter when task is started rather than forking the existing interpreter - and will pass the task_id, dag_id and configuration to the new interpreter so that it can re-initialize Airflow and all configuration in the new interpreter.

**Reply by kiranzo** [reply]
Deadline to MVP is approaching, but I'll test it when I have more time.

**Reply by kiranzo** [reply]
Hello again.

I have new problem of the same nature. 

When using [Indigo](https://github.com/epam/Indigo) chemoinformatics package (`layout()` function to be precise), DAG froze unexpectedly after one of many runs. This is really weird, because we already processed several hundred thousands of molecules before, and it never happened.

I contacted devs, and they said Indigo Python API doesn't parallelize, but may use threadsafe methods such as `lock()` and suggested rewring code to use thread local storage. 

I also tried setting `mp_start_method`  options like you suggested but to no avail. 

We skipped this one batch and continued running ETL on other molecules, and no problems enconuntered so far, only this one batch keeps freezing on this one specific molecule, while also working correctly outside of Airflow on the same Python instance.

@potiuk  Can you please suggest anything else I can try here?

**Reply by potiuk** [reply]
Not really. I do not know indigo and the fact that this one package is freezing indicates some peculiar error with it. but I have no other suggestion - maybe it tries to do soem interactive tasks and asking user? or freezez because it uses too much memory and compound with Airfflow classes overhead it reaches some limits? Really no idea here.
